### PR TITLE
[SPARK-LLAP-70]Implement against Apache Spark 2.2

### PR DESCRIPTION
**Pull request by merlintang**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by merlintang** [comment]
@bikassaha @dongjoon-hyun  you can review this later after I have done the testing.

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @merlingtang. Please let me know after the test pass.

**Comment by dongjoon-hyun** [comment]
So, the test suite is not passed here for some reasons?

**Comment by dongjoon-hyun** [comment]
Could you share the test result? Which is passed and failed?

File: patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
++import org.apache.spark.internal.Logging
++import org.apache.spark.util.Utils
++
++private[security] class HiveServer2CredentialProvider extends ServiceCredentialProvider
++  with Logging {
++
++  override def serviceName: String = "hiveserver2"
++
++  override def obtainCredentials(
++                                  hadoopConf: Configuration,
```
Comment (dongjoon-hyun) [inline patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch]: Indentation

File: patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
++        con.close()
++        con = null
++      }
++    }
++
++    None
++  }
++
++  /**
++    * Run some code as the real logged in user (which may differ from the current user, for
```
Comment (dongjoon-hyun) [inline patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch]: Indentation.

File: patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+-import org.apache.spark.sql.internal.StaticSQLConf.CATALOG_IMPLEMENTATION
++import org.apache.spark.sql.internal.StaticSQLConf.{CATALOG_IMPLEMENTATION, LLAP_ENABLED}
+ import org.apache.spark.sql.sources.BaseRelation
+ import org.apache.spark.sql.streaming._
+ import org.apache.spark.sql.types.{DataType, StructType}
+@@ -1031,9 +1031,18 @@ object SparkSession {
+   private val HIVE_SESSION_STATE_BUILDER_CLASS_NAME =
+     "org.apache.spark.sql.hive.HiveSessionStateBuilder"
+ 
++  // HDP Llap SessionStateBuilder
```
Comment (dongjoon-hyun) [inline patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch]: Remove `HDP`.

File: patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
++  private val LLAP_SESSION_STATE_BUILDER_CLASS_NAME =
++  "org.apache.spark.sql.hive.LlapSessionStateBuilder"
```
Comment (dongjoon-hyun) [inline patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch]: Indentation.

File: patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+       case "in-memory" => classOf[SessionStateBuilder].getCanonicalName
+     }
+   }
+@@ -1069,4 +1078,27 @@ object SparkSession {
+     }
+   }
+ 
++
++  /**
++    * Return true if `spark.sql.hive.llap=true` and classes can be loaded.
```
Comment (dongjoon-hyun) [inline patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch]: Indentation.

File: patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+index c8a4311..5774eca 100644
+--- a/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HiveServer2CredentialProvider.scala
++++ b/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HiveServer2CredentialProvider.scala
+@@ -87,9 +87,9 @@ private[security] class HiveServer2CredentialProvider extends ServiceCredentialP
+   }
+ 
+   /**
+-    * Run some code as the real logged in user (which may differ from the current user, for
+-    * example, when using proxying).
+-    */
```
Comment (dongjoon-hyun) [inline patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch]: Could you fix all indentation issues? Especially, this is a wrong update.

File: patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+@@ -1078,12 +1078,11 @@ object SparkSession {
+     }
+   }
+ 
+-
+   /**
+-    * Return true if `spark.sql.hive.llap=true` and classes can be loaded.
+-    * On class loading errors, it will fails.
+-    * Return false if `spark.sql.hive.llap=false`.
+-    */
```
Comment (dongjoon-hyun) [inline patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch]: Uh... The patch file seems to be wrong.

**Comment by dongjoon-hyun** [comment]
I finished my first round. @merlintang .

File: build.sbt [hunk 8]
```diff
@@ -1,6 +1,6 @@
 
 name := "spark-llap"
-version := "1.1.1-2.1"
+version := "1.1.1-2.2"
```
Comment (dongjoon-hyun) [inline build.sbt]: You need to change the versions in `assembly` together.

File: patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch (hunk 6 above)
Comment (merlintang) [inline patch/0001-SPARK-LLAP-RANGER-Integration_2.2.0.patch]: thanks, I will update all of them.

**Comment by dongjoon-hyun** [comment]
Hi, @merlintang .
It seems that it's almost done.
Please change the PR description like https://github.com/hortonworks-spark/spark-llap/issues/27 .

File: build.sbt (hunk 8 above)
Comment (dongjoon-hyun) [inline build.sbt]: Did you change assembly versions?

**Comment by dongjoon-hyun** [comment]
It's not urgent. Just I left two more comments.

**Comment by dongjoon-hyun** [comment]
FYI. https://github.com/apache/spark/pull/17710

**Comment by merlintang** [comment]
thanks for letting me know. I will keep tack this information.

**Comment by merlintang** [comment]
I will close this one, since I will open a new PR based on the latest updated branch-2.2.

**Pull request closed without merging**